### ISSUE TITLE
[HOLD for payment 2022-09-07] [$2000] IOU - User is unable to modify the IOU amount - Reported by: @Santhosh-Sellavel 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#6154]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to a conversation 
2. Click on attachment button
3. Click on request money
4. On the amount, enter 1550
5. Try to modify the amount 

## Expected Result:
User should be able to modify the amount by either using the `<` back caret from the end of the number, or highlighting any grouping of numbers and deleting them with a single keypress using the `<` caret.

## Actual Result:
User is unable to tap to highlight the numbers on the BNP

## Workaround:
Use the back caret `<` to delete/modify from the last number. 

## Expected Solution:
1. Fix the issue on all platforms by using the `selection` and `value` prop of TextInput.
2. Don't use setNativeProps as it won't be supported in [Fabric](https://github.com/facebook/react-native/issues/34113#issuecomment-1171761419) 
3. Don't use any workaround like setTimeout, setImmediate, requestAnimationFrame etc.

## Drawbacks of the expected solution:
1. On Android, Copy / Paste menu will not work. But it'll eventually be fixed by https://github.com/Expensify/App/issues/8349
2. From past experience, I'm assuming poor performance (at least on the dev build) when
    - typing quickly
     - or moving the cursor around

I believe that upgrading to the new Fabric architecture will fix both the drawbacks ([anecdotal evidence](https://expensify.slack.com/archives/C01GTK53T8Q/p1654878284799759)). 
We already have a ticket for the upgrade over here - #8503 

## Platform:

Where is this issue occurring?

- iOS
- Android

**Version Number:** 1.1.12-0

**Reproducible in staging?:** Yes 
**Reproducible in production?:** Yes 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Upwork:** https://www.upwork.com/jobs/~017b2406f1c8e307db

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/44479856/139747668-2fc1d5c5-da25-4b47-9940-2827edf67e1e.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Santhosh-Sellavel 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1635545417275400

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @deetergp (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by deetergp** [comment]
This can be external, but I want to verify that this is the bug and not the behavior we see on web/desktop.

**Comment by deetergp** [comment]
The consensus is, let's do it 👍 Setting external…

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by trjExpensify** [comment]
>The consensus is, let's do it 👍 

👋 @deetergp, can you point me to this please? I can only find this thread with no discussion: https://expensify.slack.com/archives/C01GTK53T8Q/p1635545417275400

**Comment by deetergp** [comment]
I believe [this is the thread](https://expensify.slack.com/archives/C01SKUP7QR0/p1635960613171100) you want @trjExpensify.

**Comment by trjExpensify** [comment]
Got it! So in terms of the expected behaviour then. Are we advocating for highlighting the number on press, and then tapping the caret `<` on the keyboard deletes just that selected number?

**Comment by MelvinBot** [comment]
@trjExpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by trjExpensify** [comment]
^^ bump

**Comment by MelvinBot** [comment]
@trjExpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by trjExpensify** [comment]
@deetergp can you confirm [this](https://github.com/Expensify/App/issues/6154#issuecomment-962046536) is the intended behaviour here, so I can update the OP accordingly?

**Comment by deetergp** [comment]
> Got it! So in terms of the expected behaviour then. Are we advocating for highlighting the number on press, and then tapping the caret < on the keyboard deletes just that selected number?

Yes, that's correct. They should be able to either just `<` back from the end of the number, or highlight any grouping of numbers and delete them with a single keypress.

**Comment by MelvinBot** [comment]
@trjExpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by parasharrajat** [comment]
This issue and https://github.com/Expensify/App/issues/6285 are correlated. And the solution of other will directly affect this issue. So I suggest holding it until we are done with the other one which already has some proposals.

**Comment by trjExpensify** [comment]
Cool, sounds good @parasharrajat. I'll pop this on hold and weekly for the time being.

**Comment by trjExpensify** [comment]
Still held 👍

**Comment by trjExpensify** [comment]
@parasharrajat presuming we're still held here?

**Comment by parasharrajat** [comment]
Yes. @trjExpensify

**Comment by trjExpensify** [comment]
HNY, @parasharrajat! What's the latest here?

**Comment by parasharrajat** [comment]
We should be good to go here @trjExpensify

**Comment by trjExpensify** [comment]
As in, good to close this out or proceed with moving it to Upwork? 😅

**Comment by parasharrajat** [comment]
Sorry, I meant you can post it to Upwork now. We need more work for this to work properly. Thanks. @trjExpensify

**Comment by trjExpensify** [comment]
Nice! Posted to Upwork here: https://www.upwork.com/jobs/~01d80641693bce80a7

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by zmays** [comment]
I've reproduced this bug, and believe the culprit is this code here:

```
    /**
     * Update amount with number or Backspace pressed for BigNumberPad.
     * Validate new amount with decimal number regex up to 6 digits and 2 decimal digit to enable Next button
     *
     * @param {String} key
     */
    updateAmountNumberPad(key) {
        // Backspace button is pressed
        if (key === '<' || key === 'Backspace') {
            if (this.state.amount.length > 0) {
                this.setState(prevState => ({
                    amount: prevState.amount.slice(0, -1),
                }));
            }
            return;
        }

        this.setState((prevState) => {
            const amount = `${prevState.amount}${key}`;
            return this.validateAmount(amount) ? {amount: this.stripCommaFromAmount(amount)} : prevState;
        });
    }
```

To fix I'd add logic to updateAmountNumberPad that would figure out either:
- where the cursor is, and slice from that index, or
- what numbers are highlighted, and slice all of those

**Comment by parasharrajat** [comment]
Thanks, @zmays, I agree with your plan but could you please share some technicals on How are you planning to achieve this?

> To fix I'd add logic to updateAmountNumberPad that would figure out either:

    where the cursor is, and slice from that index, or
    what numbers are highlighted, and slice all of those

**Comment by parasharrajat** [comment]
LOL, @rushatgabhane Sorry, you take it up. Didn't see the assignee.

**Comment by rushatgabhane** [comment]
Haha np 😂 
Yes, please share the details Rajat has asked on this. 

You might have to use selection and onSelectionChange props of TextInput. But it causes the cursor to jump around if not done properly. So yeah, the exact logic is important here.

**Comment by zmays** [comment]
Still got one bug (and the code is messy), but this is pretty close:

new state:
```
        this.state = {
            amount: props.selectedAmount,
            currentSelection: { start: 0, end: 0 }
        };
```

new props for TextInput
```
                        selection={this.state.currentSelection}
                        onSelectionChange={this.inputSelectionChange}
```

new methods:

```
    inputSelectionChange(event) {
        console.log(event.nativeEvent.selection)
        this.setState({ currentSelection: event.nativeEvent.selection })
    }

    cursorPositionToNewStringAndPosition(input, start, end) {
        console.log(input, start, end)
        if (input.length < 1) {
            return [input, input.length]
        }
        if (start === 0 && end === 0) {
            return [input, 0]
        }

        const inputAsArray = input.split('')

        if (start === end) {
            console.log("equals")
            return [inputAsArray.filter((e, i) => i !== start - 1).join(''), start - 1]
        } else {
            console.log("filter")
            return [inputAsArray.filter((e, i) => {
                console.log(i, start, end)
                return i <= start - 1 || i > end - 1
            }).join(''), (end - start) - 1]
        }
    }

    updateAmountNumberPad(key) {
        // Backspace button is pressed
        if (key === '<' || key === 'Backspace') {
            if (this.state.amount.length > 0) {
                const beginIndex = this.state.currentSelection.start ?? this.state.amount.length
                const endIndex = this.state.currentSelection.end ?? this.state.amount.length
                const [newAmount, newPosition] = this.cursorPositionToNewStringAndPosition(this.state.amount, beginIndex, endIndex)
                this.setState(prevState => {
                    return {
                        amount: newAmount,
                        currentSelection: { start: newPosition, end: newPosition }
                    }
                });
            }
            return;
        }

        this.setState((prevState) => {
            const amount = `${prevState.amount}${key}`;
            const valid = this.validateAmount(amount)
            if (valid) {
                const stripped = this.stripCommaFromAmount(amount)
                return {
                    amount: stripped,
                    currentSelection: { start: stripped.length, end: stripped.length }
                }
            } else {
                return prevState
            }
        });
    }

```

Summary:

- track cursor selections, set sane default if they never touch the cursor
- when backspacing create new string and set new cursor position
- when typing also set new cursor position
- if cursor is at front of the input, don't backspace

The bug I need to fix is if you make a partial selection, and delete those items, the next backspace after that does nothing

**Comment by rushatgabhane** [comment]
Hey @zmays, thanks for the details.
Apart from the bugs you've mentioned, I could find this cursor jumping bug when you type fast.

https://user-images.githubusercontent.com/29673073/148657516-1bf52d95-c748-494b-a5af-6c49ac0021aa.mp4



After you're able to find a fix for these bugs, please update your proposal.

Friendly reminder to check out [STYLE.md](https://github.com/Expensify/App/blob/main/STYLE.md). 
* Try to use a simpler logic.
* [Use underscore.js instead of native methods](https://github.com/Expensify/App/blob/main/STYLE.md#object--array-methods),
* [Use lodashGet instead of nullish coalescing operator (??)](https://github.com/Expensify/App/blob/main/STYLE.md#accessing-object-properties-and-default-values),
and yeah I'm assuming you'll make the code cleaner for a PR (redundant else, return early etc).

**Comment by rushatgabhane** [comment]
Hey @zmays, thanks for the details.
Apart from the bugs you've mentioned, I could find this cursor jumping bug when you type fast.

https://user-images.githubusercontent.com/29673073/148657516-1bf52d95-c748-494b-a5af-6c49ac0021aa.mp4



After you're able to find a fix for these bugs, please update your proposal.

Friendly reminder to check out [STYLE.md](https://github.com/Expensify/App/blob/main/STYLE.md). 
* Try to use a simpler logic.
* [Use underscore.js instead of native methods](https://github.com/Expensify/App/blob/main/STYLE.md#object--array-methods),
* [Use lodashGet instead of nullish coalescing operator (??)](https://github.com/Expensify/App/blob/main/STYLE.md#accessing-object-properties-and-default-values),
and yeah I'm assuming you'll make the code cleaner for a PR (redundant else, return early etc).

**Comment by zmays** [comment]
Spent a few hours this morning on this, and discovered that the selection binding is not working at all (i.e it is being set in state correctly, but the TextInput is ignoring it). See this issue: https://github.com/facebook/react-native/issues/29063

Here is the current state of the code: https://github.com/zmays/App/pull/1/files

The above issue references using setNativeProps, which I added to the above wip pr. This seems to partially work, in that it now will backspace from the correct place, but the cursor itself is not updating to the correct position.

I can keep trying to get this to work, but want to see what direction you think is best, given above info.

**Comment by rushatgabhane** [comment]
@zmays really appreciate your insights 😄
Found [this comment](https://github.com/facebook/react-native/issues/29063#issuecomment-814888570) on the RN issue you shared. Maybe setting selection using `inputRef.current.setNativeProps()` might help?

**Comment by zmays** [comment]
@rushatgabhane much appreciated!

I saw that comment as well, and already used it, please see code here: https://github.com/zmays/App/pull/1/files and comment from above post:

```
The above issue references using setNativeProps, which I added to the above wip pr. This seems to partially work, in that it now will backspace from the correct place, but the cursor itself is not updating to the correct position.
```

**Comment by rushatgabhane** [comment]
My bad, apologies!
Umm correct me if I'm wrong. For multiple backspaces the amount changes correctly, but the cursor stays the same even though you're setting it correctly using `setNativeProps` in `componentDidUpdate` 

This is because of RN's issue right?

**Comment by zmays** [comment]
No problem! And yes, exactly. Here is a video, showing desired behavior (in terms of what the backspace button deletes), but you will notice the visible cursor is acting weird, in that it is not in the right spot.

https://user-images.githubusercontent.com/4370615/149040797-003f923b-173d-4a02-a664-86358428b416.mov

Also this makes me believe the bug you noted before here https://github.com/Expensify/App/issues/6154#issuecomment-1008124386 about the cursor hopping around is unrelated to these code changes.

**Comment by zmays** [comment]
Okay, thanks @rushatgabhane for trying to reproduce on main.

So are you saying this task should be delayed until this https://github.com/facebook/react-native/issues/29063 issue is resolved?

what do you mean by controlled carets?

**Comment by rushatgabhane** [comment]
> So are you saying this task should be delayed until this facebook/react-native#29063 issue is resolved?

I'm not certain that it'll fix this issue. But will keep an eye on it.

> what do you mean by controlled carets?

Umm I mean controlling the cursor using `selection`.

I vote to not do anything until we can figure out a fix for the [cursor hopping bug](https://github.com/Expensify/App/issues/6154#issuecomment-1008124386), because it isn't present on `main`. 
**Or fix the issue using some alternative approach**.

cc: @puneetlath What do you think?

**Comment by K4tsuki** [comment]
#### proposal

Result (I forgot to activate `next button`):

https://user-images.githubusercontent.com/1313124/150669728-fb12acf0-8730-4645-a2b8-2ad57e6b9736.mp4

The problem is App doesn't handle text selection on `BigNumberPad` onPress. `BigNumberPad` is try to imitate user input but there is no selection handling. Also there is another bug(s) when input IOU amount in `IOUAmountPage` and inserting a value seems heavy.

On `IOUAmountPage` add reference to l`TextInputFocusable` in android or IOS. 
It might be wrapper around `FocusableTextInput`, so it is not strict on `TextInputFocusable`. 
Move amount state from `IOUAmountPage` to `TextInputFocusable` element (it could be amount or value), here to make `IOUAmountPage` lighter because re-rendering entire `IOUAmountPage` on each number input seems heavy.
 
Inside `IOUAmountPage` add reference to `TextInputFocusable` by adding a  `ref` property in `TextInputAutoWidthWithoutKeyboard` and ref value is function, this could be:

<details> <summary> setReference:  </summary>

```
setReference ( el, focusableElement ) {
     this.textInput = el;
     this.focusableEl = focusableElement;
}
```
</details>

In `FocusableTextInput` `componentDidMount` calling reference function will be `this.props.forwardedRef(this.textInput, this);`

------------------------

In `IOUAmountPage` add `onSelectionChange` property and save selection value in variable. `onSelectionChange` handling could be: `this.selectionValue = e.nativeEvent.selection;`

Modify `IOUAmountPage` BigPad press hander, could be:

<details> <summary> updateAmountNumberPad </summary>

```
updateAmountNumberPad(key) {
    const { start,  end} = this.selectionValueval;
    const focusableEl = this.focusableEl;
  	const amount = focusableEl.getValue();
    if (key === '<' || key === 'Backspace') {
        if (end == 0) {
            return;
        }
        const nextAmount =
            start == end ?
            amount.slice(0, start - 1) + amount.slice(end) :
            amount.slice(0, start) + amount.slice(end);

        if (start == end && start > 0) {
            if (!this.validateAmount(nextAmount)) {
                return;
            }
            focusbleEl.setValue(nextAmount); 
        } else {
            if (!this.validateAmount(nextAmount)) {
                return;
            }
            focusableEl.setValue(nextAmount, {  start, end: start });
        }
        return;
    }
  
    const frontPart = amount.slice(0, start);
    const endPart = amount.slice(end);
    const nextAmount = frontPart + key + endPart;

    if (!this.validateAmount(nextAmount)) {
        return;
    }

    focusableEl.setValue(nextAmount, { start: start + 1, end: start + 1}); 
}
``` 

</details> 



Here we add some functions inside `TextIputFocusable` that has been used in code above: `setValue( value, selction )`, `getValue()`.  and modify `render` function, could be:

<details> <summary> TextInputFocusable render function:  </summary>

```
render() {
  if (this.isSetSelectionActive) {
    this.isSetSelectionActive = false;
    return (
      <TextInput
        key="1"
        placeholderTextColor={themeColors.placeholderText}
        value={this.state.value}
        selection={this.selectionValue}
        ref={(el) => (this.textInput = el)}
        maxHeight={CONST.COMPOSER_MAX_HEIGHT}
        rejectResponderTermination={false}
        editable={!this.props.isDisabled}
        {...this.props}
      />
    );
  } else {
    return (
      <TextInput
        key="1"
        placeholderTextColor={themeColors.placeholderText}
        value={this.state.value}
        ref={(el) => (this.textInput = el)}
        maxHeight={CONST.COMPOSER_MAX_HEIGHT}
        rejectResponderTermination={false}
        editable={!this.props.isDisabled}
        {...this.props}
      />
    );
  }
}

```

</details>

The difference between two TextInput returned by render function is one with `selection` property. Use `Selection` property minimally (only on setting cursor or selection position) because native text input handler on user interaction in `TextInput` doesn't display context menu properly if `selection` property is used.

Then on setvalue:  

<details> <summary> setValue:  </summary>

```
function setValue(value, nextSelectionl) {
  if (nextSelection) {
    this.setSelection(nextSelection);
    window.setTimeout(() => {
      this.setState({ value });
    }, 10);
  }
  this.setState({ value });
}

```

</details>

`setTimout` here used to re-render `TextInput` so that `TextInput`  without selection property is returned in render function. 
Because using TextInput with `selection` doesn't handle context menu properly. Use  `selection` only in setting cursor or selection position.

<details> <summary> getValue():  </summary>

```
function getValue() {
  return this.state.value;
}
```

</details>

<details> <summary> setSelection:  </summary>

```
function setSelection(selection){
    this.isSetSelectionActive = true; 
    this.selectionValue = selection;
```

</details>

That is my solution or similar solution along the line.

**Comment by rushatgabhane** [comment]
I recently tested that the cursor jumping bug isn't present on release variants of the app. Feel free to update your proposals.
cc: @zmays @K4tsuki

**Comment by rushatgabhane** [comment]
@K4tsuki I find your proposal interesting, but it feels like a workaround. 

I'm gonna refer [Rajat's points](https://github.com/Expensify/App/issues/6876#issuecomment-1019030061) since both of your proposals are similiar.
>Our codebase strictly prohibits the usage of setTimeout unless this is the last resort and we don't have control over the API. So we should try to get away from it as much as possible.
In your code example, you are remounting the TextInput which is not really the React way. The same thing should be achieved either from props or the state.

**Comment by sig5** [comment]
*Proposal*
There are a few changes that can be done in order to rectify the above issue:
1. We need to add a `selection` object as the state of the `IOUAmountPage.js`. This state is used by Inputs to track the selection start and selection end points.
```
 this.state = {
            amount: props.selectedAmount,
            selection: {
                start: props.selectedAmount.length,
                end: props.selectedAmount.length,
            },
        };

```
2. The state will be passed down to our `TextInputAutoWidthWithoutKeyboard` Component.
  ```
<TextInputAutoWidthWithoutKeyboard
                        inputStyle={styles.iouAmountTextInput}
                        textStyle={styles.iouAmountText}
                        onChangeText={this.updateAmount}
                        ref={el => this.textInput = el}
                        value={formattedAmount}
                        placeholder={this.props.numberFormat(0)}
                        keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
                        selection={this.state.selection}
                        onSelectionChange={this.onSelectionChange}
                    />
```
3. The onSelectionChange  listener for our selection state can be propagated as under:
```
   onSelectionChange(e) {
        this.setState({selection: e.nativeEvent.selection});
    }

```
e.nativeEvent provides us with the user selection at the moment.

4. The following changes are mostly conditional and concerned with the amount and different positions of the caret.So, the main working logic of the `UpdateAmountNumberPad` method can be modified as follows:

```updateAmountNumberPad(key) {
        const {start, end} = this.state.selection;
        // Backspace button is pressed
        if (key === '<' || key === 'Backspace' && this.state.amount.length > 0) {
            if (end == 0) {
                return;
            }
            const amount = 
                start === end ?
                this.state.amount.slice(0, start - 1) + this.state.amount.slice(end) :
                this.state.amount.slice(0, start) + this.state.amount.slice(end);

            if (!this.validateAmount(amount)) {
                return;
            }

            if (start == end && start > 0) {
                return this.setState(prevState => ({
                    amount,
                    selection: {
                        start: prevState.selection.start - 1,
                        end: prevState.selection.end - 1,
                    }
                }));
            }

            return this.setState(prevState => ({
                amount,
                selection: {
                    start: prevState.selection.start,
                    end: prevState.selection.end,
                }
            }));
        }


        this.setState((prevState) => {
            const amount = `${prevState.amount.slice(0, start)}${key}${prevState.amount.slice(end)}`;
            return this.validateAmount(amount) ? {
                amount: this.stripCommaFromAmount(amount),
                selection: {
                    start: prevState.selection.start + 1,
                    end: prevState.selection.end + 1,
                } 
            } : prevState;
        });
    }
```
This seems to work fine on the release build for me. The cursor hopping bug occurs due to overbundled build as discussed with Rajat once in Slack.

**Comment by rushatgabhane** [comment]
Thanks for your proposal @sig5 
It has the same issue - **Copy / Paste menu is not being displayed** https://github.com/Expensify/App/issues/6876 

But I believe we can safely ignore this. Once #6876 is fixed, it should be fixed here too!


https://user-images.githubusercontent.com/29673073/150701161-6e0a7a33-f778-4656-804b-b949062c6346.mp4

**Comment by rushatgabhane** [comment]
@parasharrajat wanted your input on [my comment above](https://github.com/Expensify/App/issues/6154#issuecomment-1019584341) since you're assigned to 6876.

**Comment by rushatgabhane** [comment]
@parasharrajat could you please add your thoughts on [my comment above](https://github.com/Expensify/App/issues/6154#issuecomment-1019584341) since you're assigned to 6876.

**Comment by parasharrajat** [comment]
I haven't tested it but are you saying that copy paste menu issue is also occuring on this input when we use selection prop. @rushatgabhane 

And 6876 only handles composer input so that won't the same issue here. Can we use selection api here instead of prop?

**Comment by rushatgabhane** [comment]
> are you saying that copy paste menu issue is also occuring on this input when we use selection prop

Yes

**Comment by rushatgabhane** [comment]
Alright thanks! 
Let's see if selection api helps with the paste menu.

**Comment by K4tsuki** [comment]
Isn't Selection API related to web.
This is  native problem.

**Comment by K4tsuki** [comment]
Thank @parasharrajat 
using `textInput.setNativeProps({selection});` works.
Then conditional return in render above and setTiemout cold be removed.
This solve `remounting` and `setTimeout` in @parasharrajat  comment:
> Our codebase strictly prohibits the usage of setTimeout unless this is the last resort and we don't have control over the API. So we should try to get away from it as much as possible.
In your code example, you are remounting the TextInput which is not really the React way. The same thing should be achieved either from props or the state.

**Comment by rushatgabhane** [comment]
@K4tsuki perfect!
Can you please update your proposal to reflect the changes?

**Comment by K4tsuki** [comment]
#### proposal

My proposal is similar to my previous proposal. 
But I Will make `IOUTextInput` that wrap or above`TextInputAutoWidthWithoutKeyboard` that act like `TextInputFocusable` that I explain above for native. For web then no change, just return `TextInputAutoWidthWithoutKeyboard`.

My plan is:
Will make reference to this element in `IOUAmountPage` and move validation function inside this element and will process `PadPress` by providing this method that will process BigNumberPad pressed event on IOUAmountPPage. 
This element also save `selection` value  in variable.  Also provide `onValidityStateChange` props event that could be handled by `IOUAmountPage` to `setState` for button activation.

`setValue` using `setNativeProps` could be: 

```
function setValue(amount, formattedAmount, nextSelection) {
  this.textInput.setNativeProps({ selection: nextSelection });
  this.setState({ amout, formattedAmount });
}
```

Render method could be returning:

```
<TextInputAutoWidthWithoutKeyboard
  inputStyle={styles.iouAmountTextInput}
  textStyle={styles.iouAmountText}
  onChangeText={this.onChangeText}
  ref={this.setReference}
  value={this.state.formattedAmount}
  placeholder="0"
  keyboardType={CONST.KEYBOARD_TYPE.NUMERIC}
  onSelectionChange={this.onSelectionChange}
/>;

If you have any questions, please let me know."

**Comment by rushatgabhane** [comment]
>  I Will make IOUTextInput that wrap or aboveTextInputAutoWidthWithoutKeyboard that act like TextInputFocusable

@K4tsuki I don't think we should do that when we're trying to consolidate all the TextInputs https://github.com/Expensify/App/issues/6584

**Comment by K4tsuki** [comment]
@rushatgabhane the name can be changed, but the function is essentially the same.
Should I manage all state in `IOUAmountPage`? But I think this is not best practice, and as you can see inserting a number feels slow because render  is called in `IOUAmountPage` each time user  input a number.

From https://reactjs.org/docs/thinking-in-react.html :
>But how do you know what should be its own component? Use the same techniques for deciding if you should create a new function or object. One such technique is the single responsibility principle, that is, a component should ideally only do one thing. If it ends up growing, it should be decomposed into smaller subcomponents.

But if you want to combine the states in `IOUAmountPage`  then I will just follow you, the functions is essentially the same.

**Comment by sig5** [comment]
***Proposal***
I would like to add to my previous proposal here: https://github.com/Expensify/App/issues/6154#issuecomment-1019581224
Instead of passing down the selection as a prop, we can use the selection API to update the results in the textInput as follows:
`this.textInput.setNativeProps({selection:this.state.selection})`

https://user-images.githubusercontent.com/48352577/150905797-2de7ab15-6890-41c4-a4db-75950f5928dd.mp4

This seems to work alright with no bugs on my phone.

**Comment by rushatgabhane** [comment]
> But I think this is not best practice, and as you can see inserting a number feels slow and could be the cause of cursor hopping bug because render is called in IOUAmountPage each time user input a number

I think we should avoid workarounds in this case, performance seems good to me even on 5 year old phone.
On that note, we should probably convert `Button` to a `PureComponent` to avoid re renders.

**Comment by rushatgabhane** [comment]
🎀👀🎀 C+ reviewed 

I recommend @sig5's [proposal](https://github.com/Expensify/App/issues/6154#issuecomment-1020776843). Tough choice, but it's the first correct proposal. Rajat came up with a fix for the [drawback of the proposal](https://github.com/Expensify/App/issues/6154#issuecomment-1019593761), so I'm discounting comments after that.

cc: @puneetlath

**Comment by puneetlath** [comment]
Ok sounds good to me. One question before we move forward. 

> Instead of passing down the selection as a prop, we can use the selection API to update the results in the textInput whenever we need to update the selection state object

@sig5 why is this better than passing it as a prop?

**Comment by rushatgabhane** [comment]
> why is this better than passing it as a prop?

Afaik, passing `selection` as a prop makes it controlled which leads to [this bug (copy/paste menu doesn't show)](https://github.com/Expensify/App/issues/6154#issuecomment-1019584341 ), and poor performance which causes cursor to jump back and forth on typing fast.

**Comment by sig5** [comment]
> @puneetlath afaik, passing selection as a prop makes it controlled which leads to this bug (copy/paste menu doesn't show), and poor performance which causes cursor to jump back and forth on typing fast.

Yeah, I felt the same way about it.

**Comment by puneetlath** [comment]
Got it, that makes sense. @trjExpensify let's go forward with hiring @sig5 for this job. Thanks!

**Comment by MelvinBot** [comment]
📣 @sig5 You have been assigned to this job by @puneetlath!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by K4tsuki** [comment]
@puneetlath have you seen my [proposal](https://github.com/Expensify/App/issues/6154#issuecomment-1019435931) ? and [this](https://github.com/Expensify/App/issues/6154#issuecomment-1020692181)? 
I have said that if @rushatgabhane  doesn't want to make new component for iouamount input (as I explain this could benefit in make page feels faster as managing amount state inside IOUAmountPage could triger entire page re-render each time user input a number) I can just follow what @rushatgabhane want and I can mange the state in IOUAmountPage. the functions is the same. 

And also I first addressing `copy paste` problem in my proposal  but @sig5 doesn't address it. 
Then @parasharrajat show some clue and we both edit our proposal and basically using same @parasharrajat clue above.But it I address the problem first in my proposal.

**Comment by K4tsuki** [comment]
> I think we should avoid workarounds in this case, performance seems good to me even on 5 year old phone. On that note, we should probably convert `Button` to a `PureComponent` to avoid re renders.

I think It is not workaround, it is the problem of design.

**Comment by rushatgabhane** [comment]
@K4tsuki this is the first correct proposal that I'm taking about - https://github.com/Expensify/App/issues/6154#issuecomment-1019581224

After that, Rajat helped by suggesting to use native API for selection which I've discounted. I hope you understand my perspective 😄

**Comment by K4tsuki** [comment]
@puneetlath 

```
<TextInputAutoWidthWithoutKeyboard ..
                      selection={this.state.selection}
                      onSelectionChange={this.onSelectionChange}
                  />
```

Here he specify `selection` property and `copy paste` problem will occur.


<br />
For this:
```
 this.state = {
            selection: {
                start: props.selectedAmount.length,
                end: props.selectedAmount.length,
            },
        };
```

and

```
 onSelectionChange(e) {
        this.setState({selection: e.nativeEvent.selection});
    }

```
He save selection in `state` that will trigger re-render each time selection is changed, I save it in variable instead and it works.

<br />

```
const {start, end} = this.state.selection;
        // Backspace button is pressed
        if (key === '<' || key === 'Backspace' && this.state.amount.length > 0) {
            if (end == 0) {
                return;
            }
            const amount = 
                start === end ?
                this.state.amount.slice(0, start - 1) + this.state.amount.slice(end) :
                this.state.amount.slice(0, start) + this.state.amount.slice(end);

            if (!this.validateAmount(amount)) {
                return;
            }

            if (start == end && start > 0) {
                return this.setState(prevState => ({
                    amount,
                    selection: {
                        start: prevState.selection.start - 1,
                        end: prevState.selection.end - 1,
                    }
                }));
            }

            return this.setState(prevState => ({
                amount,
                selection: {
                    start: prevState.selection.start,
                    end: prevState.selection.end,
                }
            }));
        }


        this.setState((prevState) => {
            const amount = `${prevState.amount.slice(0, start)}${key}${prevState.amount.slice(end)}`;
            return this.validateAmount(amount) ? {
                amount: this.stripCommaFromAmount(amount),
                selection: {
                    start: prevState.selection.start + 1,
                    end: prevState.selection.end + 1,
                } 
            } : prevState;
        });
    }

```

It is basically similar to my code  above instead it is using `state`.
This:
`const {start, end} `

and this:
```
if (key === '<' || key === 'Backspace' && this.state.amount.length > 0) {
            if (end == 0) {
                return;
            }
            const amount = 
                start === end ?
                this.state.amount.slice(0, start - 1) + this.state.amount.slice(end) :
                this.state.amount.slice(0, start) + this.state.amount.slice(end);

            if (!this.validateAmount(amount)) {
                return;
            }

```
similar to my code above

**Comment by sig5** [comment]
For clarification, I am adding one more video. I have tested it rigorously and it works fine.https://user-images.githubusercontent.com/48352577/151092806-5a86f062-fc0c-47e6-bbae-73ae85683eaf.mp4

**Comment by trjExpensify** [comment]
>Got it, that makes sense. @trjExpensify let's go forward with hiring @sig5 for this job. Thanks!

@sig5, can you apply please? :)

**Comment by sig5** [comment]
Done! Thanks :D

**Comment by trjExpensify** [comment]
Reviews ongoing, Melvin. PR marked as ready for re-review again today.

**Comment by MelvinBot** [comment]
Current assignee @rushatgabhane is eligible for the Exported assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @chiragsalian (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by puneetlath** [comment]
@chiragsalian assigned you to this issue and [the PR](https://github.com/Expensify/App/pull/7433) since I'm going on sabbatical for 2 months. Thanks!

**Comment by sig5** [comment]
This issue has been a long one and it took a real lot of digging and iterations. The amount of work that has been done for this issue has been way more than what was discussed while in the proposal phase. I would request the reviewers to consider increasing the price of this job to a 1000$ which will be more apt as per the work done and time contributed for this job.
cc: @trjExpensify

**Comment by sig5** [comment]
Thanks @trjExpensify , really appreciate it!

**Comment by sig5** [comment]
Hey folks, I have been encountering some health issues for a while due to which I am unable to enough time for the issues, even after trying my best. The work in my PR's is approaching completion and interested individuals can fork my branch and continue working on it. Thanks.

**Comment by eVoloshchak** [comment]
**Proposal**
Main idea: obtain cursor/selection position from [TextInput's](https://github.com/Expensify/App/blob/ad637d3dd7dbb03cfdc6ebcb9196944c4f6165bb/src/pages/iou/steps/IOUAmountPage.js#L228) `onSelectionChange` method and then insert/delete characters at that position
Code:
1. Get cursor/selection position of the [TextInput](https://github.com/Expensify/App/blob/ad637d3dd7dbb03cfdc6ebcb9196944c4f6165bb/src/pages/iou/steps/IOUAmountPage.js#L228) and save it to the state
```javascript
onSelectionChange={(e) => this.setState({ selection: e.nativeEvent.selection })}
``` 
2. Implement method for deleting the character
```javascript
deleteNumber = (amount) => {
    if (!this.state.selection) {
        return amount.slice(0, -1)
    } else if (this.state.selection.start === this.state.selection.end) {
        return amount.substring(0, this.state.selection.start - 1) + amount.substring(this.state.selection.start, amount.length)
    } else {
        return amount.substring(0, this.state.selection.start) + amount.substring(this.state.selection.end, amount.length)
    }
}
``` 
3. And call it [here](https://github.com/Expensify/App/blob/ad637d3dd7dbb03cfdc6ebcb9196944c4f6165bb/src/pages/iou/steps/IOUAmountPage.js#L160) (when the `<` button is pressed)
```javascript
amount: this.deleteNumber(prevState.amount),
``` 
4. Implement method for inserting the character
```javascript
insertNumber = (amount, key) => {
    if (!this.state.selection) {
        return `${amount}${key}`
    } else {
        return amount.substring(0, this.state.selection.start) + key + amount.substring(this.state.selection.end);
    }
}
``` 
5. And call it [here](https://github.com/Expensify/App/blob/ad637d3dd7dbb03cfdc6ebcb9196944c4f6165bb/src/pages/iou/steps/IOUAmountPage.js#L167)
```javascript
const amount = this.insertNumber(prevState.amount, key);
``` 

https://user-images.githubusercontent.com/9059945/164984845-5f2a2fdd-0e1a-4937-8f79-20c070bdf41c.mp4

**Comment by rushatgabhane** [comment]
@sig5 that's unfortunate to hear, get well soon!

@chiragsalian @trjExpensify this issue is now open to all contributors!

**Comment by rushatgabhane** [comment]
Also, I'm sorry that I can't help with this issue. I'm gonna unassign myself because I have school stuff going on 📚

**Comment by trjExpensify** [comment]
Sorry to hear that, @sig5. Get well soon! 

@rushatgabhane @parasharrajat @chiragsalian, are we starting from scratch to source a contributor? There seems to be a lot of context here and in the PR to capture if that's the case. I'm curious how close that PR is from being completed, and whether it's more efficient to have someone who's been closer to this issue take it over the line perhaps?

**Comment by eVoloshchak** [comment]
**Proposal:**

Same as the [original PR](https://github.com/Expensify/App/pull/7433), I used uncontrolled selection on Android (and other platforms) to avoid unnecessary renders and [this](https://github.com/Expensify/App/issues/8349), while using controlled selection on iOS, because there are problems with using `setNativeProps`.
All of the changes take place in [IOUAmountPage.js](https://github.com/Expensify/App/blob/main/src/pages/iou/steps/IOUAmountPage.js)

1. Obtain [textinput's](https://github.com/Expensify/App/blob/c308f6b79bca7f871e9a1f0ebf9423470883f1b2/src/pages/iou/steps/IOUAmountPage.js#L228) selection and save it the state (iOS) or to the variable (Android)
```javascript
onSelectionChange={(e) => {
    this.selection = e.nativeEvent.selection;
    if (Platform.OS === 'ios') {
        this.setState({ selection: e.nativeEvent.selection })
    }
}}
selection={Platform.OS === 'ios' ? this.state.selection : undefined}
```
2. Delete a symbol. We need to do two things: delete symbol and move caret to the left. First we imlement method for setting the caret position:
```javascript
setCaretPosition(position) {
    if (position < 0) {
        position = 0;
    }
    this.selection = { start: position, end: position };
    if (Platform.OS === 'ios') {
        this.setState({ selection: this.selection });
    } else {
        this.textInput.setNativeProps({ selection: this.selection });
    }
}
```
And then update the updateAmountNumberPad on [this](https://github.com/Expensify/App/blob/c308f6b79bca7f871e9a1f0ebf9423470883f1b2/src/pages/iou/steps/IOUAmountPage.js#L159) line
```javascript
this.setState(prevState => {
    const amount = this.deleteSymbol(prevState.amount);
    let cursorPosition = this.selection.start; // if user has highlighted at least one symbol, selection.start is automatically shifted to the left
    if (this.selection.start === this.selection.end) {
        cursorPosition = cursorPosition - 1; 
    }
    this.setCaretPosition(cursorPosition);
    return { amount };
});
```
3. Insert a symbol. Just insert symbol and move caret to the right. Change [this](https://github.com/Expensify/App/blob/c308f6b79bca7f871e9a1f0ebf9423470883f1b2/src/pages/iou/steps/IOUAmountPage.js#L167) line to the following:
```javascript
  const amount = prevState.amount.substring(0, this.selection.start) + key + prevState.amount.substring(this.selection.end);
  const amountIsValid = this.validateAmount(amount)
  if (amountIsValid) {
      this.setCaretPosition(this.selection.start + 1);
  } 
  return amountIsValid ? {amount: this.stripCommaFromAmount(amount)} : prevState;
```
4. If user pasted text into TextInput. There is an [issue](https://github.com/Expensify/App/pull/7433#issuecomment-1094363987) where caret position is off when user is pasting something into textInput in the middle of text, so let's address it. Essentially, we need to move cursor to the end of the pasted text. For that we need to know the length of text the user has pasted. If we have previous string, current string and a selection, we can obtain it like this:
```javascript
 getPastedText(prevString, newString) {
      const beforeSelection = prevString.substring(0, this.selection.start);
      const afterSelection = prevString.substring(this.selection.end, prevString.length);
      return newString.replace(beforeSelection, '').replace(afterSelection, '');
  }
```
And then call it in updateAmount method:
```javascript
  updateAmount(text) {
      this.setState((prevState) => {
          const amount = this.replaceAllDigits(text, this.props.fromLocaleDigit);
          const amountIsValid = this.validateAmount(amount)
          if (amountIsValid) {
              const pastedText = this.getPastedText(prevState.amount, this.stripCommaFromAmount(amount))
              this.setCaretPosition(this.selection.start + pastedText.length);
          }
          return amountIsValid
              ? {amount: this.stripCommaFromAmount(amount)}
              : prevState;
      });
  }
```
5. And lastly, I encountered an issue with iOS, where if you set `value` and `selection` simultaneously via setState, they don't get updated. I'm not sure how to resolve it, except for wrapping `setState` in `setCaretPosition` into `setTimeout`. I know this is generally avoided, but I couldn't find other was to resolve this issue, so I'm proposing it here as a [potential solution](https://github.com/Expensify/App/pull/7433#issuecomment-1040293555).

**The whole diff can be found [here](https://github.com/Expensify/App/compare/main...eVoloshchak:eVoloshchak_iouAmountPrototype#diff-6afafc66fbd449a930a6426cf437d635ade637982eba83b27fe5e04cf39590deL159).**


**Android**


https://user-images.githubusercontent.com/9059945/165144253-7ba353aa-9e1a-49b2-924e-a8a938a6d272.mp4



**mWeb**

https://user-images.githubusercontent.com/9059945/165144194-58c37ca6-941d-48d1-9ae2-14590b039a14.mp4


**Web**


https://user-images.githubusercontent.com/9059945/165144605-5d99be86-8184-41da-92c2-6136048a8393.mp4

**iOS**

https://user-images.githubusercontent.com/9059945/165145373-233b6e8f-cb8a-47a5-baa3-c856b11b6f4f.mp4

**Comment by rushatgabhane** [comment]
> are we starting from scratch to source a contributor? There seems to be a lot of context here and in the PR to capture if that's the case. I'm curious how close that PR is from being completed, and whether it's more efficient to have someone who's been closer to this issue take it over the line perhaps?

The PR was like >95% complete. But we kinda hit a hard wall here https://github.com/Expensify/App/pull/7433#pullrequestreview-923013569

It seems to be a react native issue imo. Regarding the efficiency - I'm sure any C+ assigned to this issue can catch-up after spending ~30 mins reading the PR.

**Comment by trjExpensify** [comment]
Cool, @Santhosh-Sellavel do you want to let us know what you think after you catch up on the PR? @eVoloshchak proposal vids looks pretty promising! :)

**Comment by Santhosh-Sellavel** [comment]
Sorry for the delay. Looking into this now.

**Comment by Santhosh-Sellavel** [comment]
@eVoloshchak 
The proposal looks good as it is mostly based on the existing PR.

>All of the changes take place in [IOUAmountPage.js](https://github.com/Expensify/App/blob/main/src/pages/iou/steps/IOUAmountPage.js)

Read our [Philosophy point 5](https://github.com/Expensify/App#philosophy) we don’t encourage using `Platform.OS`. So please post an updated proposal.

cc: @chiragsalian

**Comment by eVoloshchak** [comment]
@Santhosh-Sellavel 

**Updated proposal:**
I've separated [TextInput](https://github.com/Expensify/App/blob/3faa6a77b076f6e6547340b93a2298042cf36266/src/pages/iou/steps/IOUAmountPage.js#L239) into platform-specific files.
Updated diff can be found [here](https://github.com/Expensify/App/compare/main...eVoloshchak:eVoloshchak_iouAmountPrototype2).

[IOUAmountInput/index.ios.js](https://github.com/Expensify/App/compare/main...eVoloshchak:eVoloshchak_iouAmountPrototype2#diff-c6cf31b8ea837df68c36960ab65b5e2038440f38bc475b1d1d0b6c530c343ed0):
```javascript
class IOUAmountInput extends React.Component {
    constructor(props) {
        super(props);

        this.setCaretPosition = this.setCaretPosition.bind(this);
        this.focus = this.focus.bind(this);

        this.state = {
            selection: {start: 0, end: 0},
        };
    }

    setCaretPosition(position) {
        setTimeout(() => this.setState({selection: {start: position, end: position}}), 1);
    }

    focus() {
        this.textInput.focus();
    }

    render() {
        return (
            <TextInput
                ref={el => this.textInput = el}
                onSelectionChange={(e) => {
                    this.setState({selection: e.nativeEvent.selection});
                    this.props.onSelectionChange(e);
                }}
                selection={this.state.selection}
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...(_.omit(this.props, 'onSelectionChange'))}
            />
        );
    }
}
```

[IOUAmountInput/index.js](https://github.com/Expensify/App/compare/main...eVoloshchak:eVoloshchak_iouAmountPrototype2#diff-6738cab8ceb5afd825a7e897e422a29d73ff96119a11cd3e6972b0edefeb2828):
```javascript
class IOUAmountInput extends React.Component {
    constructor(props) {
        super(props);

        this.setCaretPosition = this.setCaretPosition.bind(this);
        this.focus = this.focus.bind(this);
    }

    setCaretPosition(position) {
        this.textInput.setNativeProps({selection: {start: position, end: position}});
    }

    focus() {
        this.textInput.focus();
    }

    render() {
        return (
            <TextInput
                ref={el => this.textInput = el}
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...this.props}
            />
        );
    }
}
```

**Comment by Santhosh-Sellavel** [comment]
@eVoloshchak 

Getting this weird issue:

https://user-images.githubusercontent.com/85645967/167206402-7b13e518-a750-4f25-a3d8-a10f18d1490d.mp4

**Comment by eVoloshchak** [comment]
> @eVoloshchak
> 
> Getting this weird issue, while deleting the cursor moves to start!

Sorry, I made a [typo](https://github.com/Expensify/App/commit/0f43650cb5c3c011ac6411d85e2562a77a50b4d4), now it's resolved. Same [diff](https://github.com/Expensify/App/compare/main...eVoloshchak:eVoloshchak_iouAmountPrototype2)

**Comment by Santhosh-Sellavel** [comment]
@eVoloshchak 
Still, there are some cursor issues, check out this I am trying to select a portion trying to paste 8(I tried with double digits too). One time cursor moves to different places one time it was to the end, and one time it was at the right place.

https://user-images.githubusercontent.com/85645967/167211388-f8f91da8-16ac-498e-938a-8b7395809479.mp4

**Comment by eVoloshchak** [comment]
@Santhosh-Sellavel , [updated](https://github.com/Expensify/App/compare/main...eVoloshchak:eVoloshchak_iouAmountPrototype2)

**Comment by Santhosh-Sellavel** [comment]
Can you explain, what went wrong what changes did you make?

**Comment by eVoloshchak** [comment]
> Can you explain, what went wrong what changes did you make?

Here's the [commit](https://github.com/Expensify/App/commit/de3cee81192da2ba280b8575a1e02ed2342827fd) with the changes. As you can see, previously I was using **completely** wrong logic for getting the pasted text, that's why cursor was moving wrong in some cases

**Comment by Santhosh-Sellavel** [comment]
The proposal looks good, but it's not working well on mWeb in android & iOS.

**Comment by eVoloshchak** [comment]
> I did some basic testing only things broken in Android Chrome & mWeb Safari iOS. Selection doesn’t work properly.

Here's what I found on mWeb
1. On Android end of selection isn't visible sometimes
![image](https://user-images.githubusercontent.com/9059945/168294475-abdf2416-4dd9-49a1-a9ce-25708b6de0b4.png)
However:

- This problem seems to be Chrome-specific, I couldn't reproduce it using Firefox or Opera
- It's present in current production build, so it wasn't introduced by these changes
2. An issue that was present on both iOS and Android. If you select some text, then touch empty space on screen and then start typing using keypad, the caret position is wrong (it essentially maintains the selection that it had before input lost focus). 

https://user-images.githubusercontent.com/9059945/168296352-8da8a3bd-f177-473f-ad88-fbfcaf9c1a29.mp4


To resolve that we need to do the following:
in [IOUAmountInput/index.js](https://github.com/eVoloshchak/App/blob/eVoloshchak_iouAmountPrototype2/src/pages/iou/steps/IOUAmountInput/index.js)  file (that is used only for web) add onBlur handler to the TextInput. 
```javascript
  ref={el => this.textInput = el}
  onBlur={(e) => {
      // If user pressed not on the keypad - move caret to the end
      if (!findNodeHandle(e.relatedTarget)) {
          const amountLength = this.textInput.value.length;
          const selection = {start: amountLength, end: amountLength};
          this.props.onSelectionChange({nativeEvent: {selection}});
      }
      this.props.onBlur(e);
  }}
```
Result:

https://user-images.githubusercontent.com/9059945/168296504-07a87440-bb37-43f4-adab-1922561c4049.mp4

3. In Safari, when you try to paste into an input that already contain's at leas one digit, nothing happens. However, when the input is empy, pasting works just fine. This seems strange to me, because even TextInputs onChangeText isn't called, I'm thinking maybe it's emulator issues. Do you by any chance have a physical iOS device to check this?

Besides those issues, I couldn't find any. Could you please specify which ones you were encountering?

[Updated diff](https://github.com/Expensify/App/compare/main...eVoloshchak:eVoloshchak_iouAmountPrototype2)

**Comment by trjExpensify** [comment]
Bump @Santhosh-Sellavel @chiragsalian to review the above?

**Comment by Santhosh-Sellavel** [comment]
@eVoloshchak I faced all those issues mentioned [above](https://github.com/Expensify/App/issues/6154#issuecomment-1126074886)!
Sorry, I see there was a fix too, but I missed that.

> It's present in the current production build, so it wasn't introduced by these changes

I believe those issues are well within the scope of this issue, So I would expect a fix in this issue itself. But I'll leave it to @chiragsalian to make a final call.

So what's yet to be @eVoloshchak from the  [above](https://github.com/Expensify/App/issues/6154#issuecomment-1126074886)-mentioned issues!

cc: @chiragsalian  @trjExpensify

**Comment by Santhosh-Sellavel** [comment]
@eVoloshchak 
One more question in `IOUAmountInput` component `ios` platform code `setCaretPosition` method uses `setTimeout` pattern which we generally avoid.

```js
  setCaretPosition(position) {
        setTimeout(() => this.setState({selection: {start: position, end: position}}), 1);
    }
```

Why the `setTimeout` is necessary here?

**Comment by rushatgabhane** [comment]
@Santhosh-Sellavel this will be funny if it works, maybe increasing the width by 2: `layout.width + 2`.
Fixes the [mWeb bug](https://github.com/Expensify/App/issues/6154#issuecomment-1126074886)? 

https://github.com/Expensify/App/blob/db7a1b751d9a34cac51dffd54e18907d8ecb8dd9/src/components/TextInput/BaseTextInput.js#L307

**Comment by Christinadobrzyn** [comment]
Tracking this issue to make sure I follow up with singhalsakar@gmail.com's email (sig5) based on this [convo](https://expensify.slack.com/archives/C02NK2DQWUX/p1653458528794299) about his PR & the completed job.

**Comment by Santhosh-Sellavel** [comment]
I'm going on OOO on now. Let's assign it back to @rushatgabhane if that's okay with them!
As of now, we are waiting for proposal updates/ a new proposal thanks!

**Comment by trjExpensify** [comment]
@rushatgabhane - I've assigned this back to you now that I'm back. Can you provide a quick recap of what's outstanding to be figured out here for clarity?

**Comment by rushatgabhane** [comment]
>  I've assigned this back to you

@trjExpensify thank you! 

> @rushatgabhane - Can you provide a quick recap of what's outstanding to be figured out here for clarity?

So.. we're waiting for proposals. 

We aren't looking for workaround solutions that use `setTimeout`, or platform specific code for handling `selection`.
Also, I'd like for all `react-native` bugs (if any) to be merged in https://github.com/Expensify/react-native

**Comment by rushatgabhane** [comment]
@trjExpensify I think we should remove the `reviewing` label and double the issue for more 👀

**Comment by rushatgabhane** [comment]
@trjExpensify gently bump ^^

**Comment by trjExpensify** [comment]
Ah yes, sorry about that. I caught up with @chiragsalian about this issue as it's a good 100+ comments. It's pretty lengthy, we've had some great investigations so far, but there's quite a bit of context scattered throughout the issue. What would be great is if the C+ could update the OP with the latest information on where we're at, what solutions we're avoiding and any suggestions on how best to approach the problem for whoever volunteers next. 

After which, I'll go ahead and x2 on Upwork and repost the job. In the meantime, I've removed the `reviewing` label.

**Comment by rushatgabhane** [comment]
@trjExpensify that's a great idea!

Using controlled selection causes a re-render each time we press a key or move the cursor around. 
This results in terrible performance and incorrect cursor postion when 
- typing quickly 
- or moving the cursor around

### Deliverables
- Fix the issue on all platforms _without_ using the `selection` prop. Use `setNativeProps` to set the `selection` instead.
- Use `setNativeProps` to set the amount.
- Don't use any workaround like `setTimeout`, `setImmediate`, `requestAnimationFrame` etc.

<hr />

iOS bug that you'll encounter and need to fix when using `setNativeProps` to set the amount and selection - 
"Cursor doesn't move when appending amount"

Steps
1. Go to https://snack.expo.dev/@rushat200/d2a99e
2. Click "Append random amount" button 5 times.

Expected: Cursor is at index 5.
Actual: Cursor is at index 1.

Observations
- This bug is NOT present on Android.
- This bug is NOT reproducible when using `setTimeout` on iOS, but it's only a workaround.

https://user-images.githubusercontent.com/29673073/175847402-0eab8e14-e082-4805-bbc7-c4d0a12804a9.mov


cc: @chiragsalian can you please see if I make sense, and update the OP because I don't have the permissions.

**Comment by trjExpensify** [comment]
Thanks, @rushatgabhane! ❤️ I'll await input from @chiragsalian and then have the OP updated accordingly.

**Comment by chiragsalian** [comment]
I like the deliverables section. Its very clear and concise. I don't quite get the test steps. Just to confirm, in the original OP i see,
> User is unable to tap to highlight the numbers on the BNP

So on iOS (simulator) i'm able to tap to highlight all the numbers. So is this issue still correct or is the issue that we cannot highlight a group of numbers or cannot modify highlighted numbers?

**Comment by rushatgabhane** [comment]
>  in the original OP i see

@chiragsalian ah my bad I should have added some context. Those steps have nothing to do with the OP. 

In this snack (https://snack.expo.dev/@rushat200/d2a99e), I implemented the [deliverables](https://github.com/Expensify/App/issues/6154#issuecomment-1166761257) to set selection and amount using `setNativeProps`.

```
appendAmount() {
    this.selection.start = this.selection.start + 1;
    this.selection.end = this.selection.end + 1;
    this.amount = this.amount.concat(someOneDigitNumber);

    this.textInput.setNativeProps({ text: this.amount });
    this.textInput.setNativeProps({ selection: this.selection });
}
```

This append function's expected result should be to append a 1 digit number, and move the cursor one step forward.
Makes sense?

The problem I'm highlighting is that on iOS, the cursor doesn't move one step forward!!
`setTimeout` workaround solves it, and the same code works fine on Android.

**Comment by trjExpensify** [comment]
Okay, excellent. So what are we doing here exactly? 

Adding this as a new section? 

### Expected Solution
1. Fix this react native bug https://github.com/facebook/react-native/issues/34113
   - Post a proposal on this issue. 
   - If proposal is accepted, submit a PR to `react-native` for the same.
3. Fix the issue on all platforms _without_ passing `selection` as a prop to `TextInput`. Use `setNativeProps` to set the `selection` instead.
    - `textInput.setNativeProps({ selection: this.selection })`
4. Use `setNativeProps` to set the amount.  
    - `textInput.setNativeProps({ text: this.amount })`
5. Don't use any workaround like `setTimeout`, `setImmediate`, `requestAnimationFrame` etc.

Can someone confirm where this one is supposed to go/replace: 

>Using controlled selection causes a re-render each time we press a key or move the cursor around. 
>This results in terrible performance and incorrect cursor postion when 
> - typing quickly 
> - or moving the cursor around

**Comment by rushatgabhane** [comment]
> Can someone confirm where this one is supposed to go/replace:

@trjExpensify something like this is fine

### Expected Solution 

Using controlled selection causes...

1. Fix this react native bug..
....

**Comment by rushatgabhane** [comment]
> setNativeProps will not be supported in the new "Fabric" architecture, and should be considered deprecated
https://reactnative.dev/docs/new-architecture-library-intro#migrating-off-setnativeprops

Got a reply from a maintainer that setNativeProps will be deprecated in the Fabric architecture.

@chiragsalian @trjExpensify :nevermind:
we're throwing the [expected solution](https://github.com/Expensify/App/issues/6154#issuecomment-1166761257) out of the window

**Comment by rushatgabhane** [comment]
Open to new proposals.

### Expected Solution
1. Fix the issue on all platforms by using the `selection` and `value` prop of TextInput.
2. Don't use any workaround like setTimeout, setImmediate, requestAnimationFrame etc.

### Drawbacks
1. On Android, Copy / Paste menu will not work. But it'll eventually be fixed by https://github.com/Expensify/App/issues/8349
2. From past experience, I'm assuming poor performance (at least on the dev build) when
    - typing quickly
     - or moving the cursor around

**Comment by trjExpensify** [comment]
@chiragsalian, bump on that sanity check above.

**Comment by eVoloshchak** [comment]
**Proposal:**

1. First of all, we need to keep track of [TextInput's](https://github.com/Expensify/App/blob/b9a126b07a324212863d75073a18508d6fca9159/src/pages/iou/steps/IOUAmountPage.js#L218) selection and store it in state
``` javascript
selection={this.state.selection}
onSelectionChange={e => this.setState({ selection: e.nativeEvent.selection })}
```
And pass these props through `TextInputWithCurrencySymbol` > `AmountTextInput`

2. For the case when user presses `<` (delete) on onscreen keyboard, we need to change the [updateAmountNumberPad](https://github.com/Expensify/App/blob/b9a126b07a324212863d75073a18508d6fca9159/src/pages/iou/steps/IOUAmountPage.js#L142) method 
```javascript
updateAmountNumberPad(key) {
    // Backspace button is pressed
    if (key === '<' || key === 'Backspace') {
        if (this.state.amount.length > 0) {
            this.setState((prevState) => {
                const amount = this.deleteSymbol(prevState.amount, prevState.selection);
                let cursorPosition = prevState.selection.start; // if user has highlighted at least one symbol, selection.start is automatically shifted to the left
                if (prevState.selection.start === prevState.selection.end) {
                    cursorPosition -= 1;
                }

                const selection = {start: cursorPosition, end: cursorPosition};
                return {amount, selection};
            });
        }
        return;
    }
```
Where `this.deleteSymbol()` looks like this:
```javascript
deleteSymbol(amount, selection) {
    if (selection.start === selection.end) {
        return amount.substring(0, selection.start - 1) + amount.substring(selection.start, amount.length);
    }
    return amount.substring(0, selection.start) + amount.substring(selection.end, amount.length);
}
```
3. If user pressed any other key on onscreen keyboard, the bottom half of `updateAmountNumberPad`
```javascript
this.setState((prevState) => {
    const amount = prevState.amount.substring(0, prevState.selection.start) + key + prevState.amount.substring(prevState.selection.end);
    const amountIsValid = this.validateAmount(amount);
    if (amountIsValid) {
        const selection = {start: prevState.selection.start + 1, end: prevState.selection.start + 1};
        return {amount: this.stripCommaFromAmount(amount), selection}
    }
    return prevState;
});
```
4. If input's value updated by an external keyboard or if user has pasted into the input, we need to update the [updateAmount](https://github.com/Expensify/App/blob/b9a126b07a324212863d75073a18508d6fca9159/src/pages/iou/steps/IOUAmountPage.js#L165) method
```javascript
    updateAmount(text) {
        this.setState((prevState) => {
            const amount = this.stripCommaFromAmount(this.replaceAllDigits(text, this.props.fromLocaleDigit));
            const amountIsValid = this.validateAmount(amount);
            if (amountIsValid) {
                const pastedText = this.getPastedText(prevState.amount, amount, prevState.selection);
                let cursorPosition = prevState.selection.start + pastedText.length;
                if (!pastedText && prevState.selection.start === prevState.selection.end) {
                    cursorPosition -= 1;
                }
                const selection = {start: cursorPosition, end: cursorPosition};
                return {amount, selection}
            }
            return prevState;
        });
    }
```
Where `this.getPastedText()` looks like this:
```javascript
getPastedText(prevString, newString, selection) {
    let pastedText = newString.slice(selection.start);
    if (prevString.length - selection.end) {
        pastedText = pastedText.slice(0, (prevString.length - selection.end) * -1);
    }
    return pastedText;
}
```
5. At this point everything works, but we need to fix a couple of issues. The first one i [talked about earlier](https://github.com/Expensify/App/issues/6154#issuecomment-1126074886), it is present in `mWeb` only
> If you select some text, then touch empty space on screen and then start typing using keypad, the caret position is wrong (it essentially maintains the selection that it had before input lost focus).

To resolve this, we need to check if user has touched the blank screen instead of onscreen keyboard, and if they did, move caret to the end of the input. Add `onBlur` handler for [TextInputWithCurrencySymbol](https://github.com/Expensify/App/blob/b9a126b07a324212863d75073a18508d6fca9159/src/pages/iou/steps/IOUAmountPage.js#L218) 
```javascript
onBlur={(e) => {
    // If user pressed on empty screen - move cursor to the end
    if (!findNodeHandle(e.relatedTarget)) {
        const cursorPosition = this.state.amount.length;
        this.setState({ selection: {start: cursorPosition, end: cursorPosition} })
    }
}}
```

6. This problem was also observed in [this](https://github.com/Expensify/App/pull/7433/files) PR
> iOS has incorrect selection(caret) position when value or selection is directly manipulated.

Meaning, when we have a controlled `TextInput` with `value` prop, we still set i'ts value via `setNativeProps` on every update
https://github.com/Expensify/App/blob/b9a126b07a324212863d75073a18508d6fca9159/src/components/TextInput/BaseTextInput.js#L73

To resolve that, we need to add `isValueControlled` prop to BaseTextInput, call `setNativeProps` only if it's `false`
```javascript
  if (!this.props.isValueControlled) {
      this.input.setNativeProps({text: inputValue});
  }
```
Pass `value` prop to [RNTextInput](https://github.com/Expensify/App/blob/b9a126b07a324212863d75073a18508d6fca9159/src/components/TextInput/BaseTextInput.js#L248)
```javascript
value={this.props.isValueControlled ? this.props.value : undefined}
```
And add `isValueControlled={true}` to [AmountTextInput](https://github.com/Expensify/App/blob/main/src/components/AmountTextInput.js)

**Results:**
Web:


https://user-images.githubusercontent.com/9059945/178748453-3e99d017-b8b0-4ebc-924d-1f3792a546b8.mp4






mWeb:

https://user-images.githubusercontent.com/9059945/178747348-04f3a04f-bc90-4b70-b92e-bc7187c9c050.mp4



Android:

https://user-images.githubusercontent.com/9059945/178746418-b61c3e30-b5d0-45ee-9057-f12e9d11bcff.mp4

Note: there is a flickering issue with selection on Android in DEV mode, so this video was shot with the production build

iOS:

https://user-images.githubusercontent.com/9059945/178746811-9574c96d-7f75-48e2-905e-057d55062cfb.mp4

Note: I'm unable to paste in iOS simulator due to [this](https://developer.apple.com/forums/thread/682395) bug

**Comment by trjExpensify** [comment]
OP is updated. Issue on Upwork here, bumped to $2k: https://www.upwork.com/jobs/~017b2406f1c8e307db

**Comment by eVoloshchak** [comment]
Updated my [proposal](https://github.com/Expensify/App/issues/6154#issuecomment-1183245737) according to the new requirements of not using `setNativeProps` and latest updates to `IOUAmountPage`

**Comment by chiragsalian** [comment]
Thanks for the proposal. @rushatgabhane, will you be able to do the initial review.

**Comment by rushatgabhane** [comment]
> To resolve this, we need to check if user has touched the blank screen instead of onscreen keyboard, and if they did, move caret to the end of the input. Add onBlur handler for

@eVoloshchak I think we shouldn't fix it.

**Comment by rushatgabhane** [comment]
> mWeb Safari - On pasting in middle of some amount, the cursor goes to end of amount. 

`|` denotes cursor position

Expected: 123pastedNumber`|`789

Actual: 123pastedNumber789`|`

@eVoloshchak how would you fix this bug?

**Comment by eVoloshchak** [comment]
> @eVoloshchak how would you fix this iOS mWeb bug?

I've tried running on a real device, cannot reproduce this bug.

https://user-images.githubusercontent.com/9059945/180599821-a205ed7e-29e4-4899-add7-44f395d565e4.mp4

In https://github.com/Expensify/App/pull/7433 selection logic for `mWeb` (and everything except `iOS` for that matter) was handled via `setNativeProps`, and since we use controlled selection for all platforms, this bug doesn't appear

>@eVoloshchak I think we shouldn't fix it.

No problem. This is a rare edge case, I agree

**Comment by rushatgabhane** [comment]
@eVoloshchak's proposal is solid! 

🎀👀🎀 C+ Reviewed @chiragsalian

**Comment by chiragsalian** [comment]
Proposal LGTM. @eVoloshchak feel free to create the PR.

**Comment by trjExpensify** [comment]
Awesome, sent offer on Upwork @eVoloshchak.

**Comment by eVoloshchak** [comment]
I'm waiting on https://github.com/Expensify/App/pull/9800 to be reverted, after that I'll submit a PR

**Comment by rushatgabhane** [comment]
under review

**Comment by Christinadobrzyn** [comment]
@chiragsalian, @rushatgabhane, & @trjExpensify - leaving a note here to evaluate @sig5 contributions to this job for payment when the PR is deployed to production and we're ready for payment. Convo related to this here https://expensify.slack.com/archives/C02NK2DQWUX/p1653458528794299

**Comment by trjExpensify** [comment]
Not sure why Melv is pinging us for this. https://github.com/Expensify/App/pull/10147#issuecomment-1207284244 - awaiting a review from @chiragsalian at this point.

**Comment by trjExpensify** [comment]
Ah, I misinterpreted this then.. my bad!  

>@chiragsalian feel free to jump in for a review because we're 99% there.

**Comment by trjExpensify** [comment]
@rushatgabhane, looks like you ran into a snag here with Storybook, right? https://github.com/Expensify/App/pull/10147#pullrequestreview-1074796914

**Comment by trjExpensify** [comment]
Happy Monday, Melvin! I think we're just waiting on @chiragsalian to review the PR now.

**Comment by trjExpensify** [comment]
👋 @chiragsalian @rushatgabhane I'm back from OOO. Confirming the breakdown in Slack.

**Comment by Christinadobrzyn** [comment]
I followed up with @sig5 about compensation in our email thread.

**Comment by trjExpensify** [comment]
@Santhosh-Sellavel - sent a contract for reporting 
@rushatgabhane - sent a contract for C+ 
@eVoloshchak - settled up for the fix.

**Comment by trjExpensify** [comment]
@rushatgabhane - settled up!

**Comment by trjExpensify** [comment]
@Santhosh-Sellavel - settled up. Thanks everyone! 🎉

**Comment by chiragsalian** [comment]
No idea what regression this was linked to. I don't see any relevant regression posted related to this issue. Let me know if I'm missing something.

**Comment by trjExpensify** [comment]
It's the first link in that message: https://github.com/Expensify/App/issues/10693#issuecomment-1267090302 CC: @parasharrajat
